File: frontend/src/types/majorHazard.ts
// 重大危险源类型定义。
//
// 注意：后端 Decimal 字段（临界量、q、β、s、R、α）经 FastAPI 序列化后是【字符串】
// （保精度，临界量有 0.3 / 0.75 这类值）。展示前统一用 utils/majorHazardFormat 的
// toNumber / formatQty 转换，不要直接参与算术。

export interface MajorHazardUnit {
  id: string;
  enterprise_id: string;
  name: string;
  unit_type: "production" | "storage";
  address?: string | null;
  department?: string | null;
  responsible_person?: string | null;
  responsible_phone?: string | null;
  risk_object_id?: string | null;
  is_active: boolean;
  created_at?: string | null;
}

export interface MajorHazardUnitPayload {
  name: string;
  unit_type: "production" | "storage";
  boundary_desc?: string | null;
  floor_id?: string | null;
  polygon?: Record<string, unknown> | null;
  address?: string | null;
  longitude?: number | null;
  latitude?: number | null;
  department?: string | null;
  responsible_person?: string | null;
  responsible_phone?: string | null;
  risk_object_id?: string | null;
  is_active?: boolean;
}

export interface MajorHazardUnitChemical {
  id: string;
  unit_id: string;
  chemical_id?: string | null;
  chemical_name: string;
  physical_state?: string | null;
  storage_location?: string | null;
  q_design_max: string;
  q_actual?: string | null;
  critical_quantity_t: string;
  beta: string;
  beta_source: "table3" | "table4" | "manual";
}

/** 提交品种清单时的载荷（id / unit_id 由后端管理）。 */
export type MajorHazardUnitChemicalPayload = Omit<
  MajorHazardUnitChemical,
  "id" | "unit_id"
>;

export interface MajorHazardCalculation {
  id: string;
  seq: number;
  s_value: string;
  r_value: string;
  alpha: string;
  exposed_population: number;
  is_major_hazard: boolean;
  level?: string | null;
  formula_version: string;
  calculated_at?: string | null;
}

export interface PreviewChemicalItem {
  name: string;
  q: number;
  Q: number;
  beta: number;
  q_over_Q: number;
  beta_times_q_over_Q: number;
}

/** 预览与固化共用同一形状（预览不含 seq；固化才有）。 */
export interface MajorHazardPreviewResult {
  seq?: number;
  s_value: string | number;
  r_value: string | number;
  alpha: string | number;
  exposed_population: number;
  is_major_hazard: boolean;
  level?: string | null;
  formula_version: string;
  chemicals: PreviewChemicalItem[];
}

export interface CriticalQuantity {
  chemical_name: string;
  alias?: string | null;
  cas_no?: string | null;
  critical_t?: string | null;
  critical_note?: string | null;
  table_no: string;
  source_page?: number | null;
}

export interface MajorHazardRecord {
  id: string;
  unit_id: string;
  enterprise_id: string;
  hazard_code?: string | null;
  filing_status: string;
  filing_no?: string | null;
  filing_date?: string | null;
  chief_name?: string | null;
  chief_post?: string | null;
  chief_phone?: string | null;
  tech_name?: string | null;
  tech_post?: string | null;
  tech_phone?: string | null;
  oper_name?: string | null;
  oper_post?: string | null;
  oper_phone?: string | null;
  attachments: Record<string, unknown>;
  completeness: Record<string, unknown>;
}

export interface EvidenceItem {
  id: string;
  regulation_id?: string | null;
  article_anchor: string;
  relation: string;
  note?: string | null;
}

File: backend/app/schemas/major_hazard.py
"""重大危险源 API 出入参。"""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


class UnitIn(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    unit_type: str = Field(pattern="^(production|storage)$")
    boundary_desc: Optional[str] = None
    floor_id: Optional[str] = None
    polygon: Optional[dict] = None
    address: Optional[str] = None
    longitude: Optional[float] = None
    latitude: Optional[float] = None
    department: Optional[str] = None
    responsible_person: Optional[str] = None
    responsible_phone: Optional[str] = None
    risk_object_id: Optional[str] = None
    is_active: bool = True


class UnitOut(BaseModel):
    id: str
    enterprise_id: str
    name: str
    unit_type: str
    address: Optional[str] = None
    department: Optional[str] = None
    responsible_person: Optional[str] = None
    responsible_phone: Optional[str] = None
    risk_object_id: Optional[str] = None
    is_active: bool
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class UnitChemicalIn(BaseModel):
    chemical_name: str = Field(min_length=1, max_length=500)
    chemical_id: Optional[str] = None
    physical_state: Optional[str] = None
    storage_location: Optional[str] = None
    q_design_max: Decimal = Field(ge=0)
    q_actual: Optional[Decimal] = Field(default=None, ge=0)
    critical_quantity_t: Decimal = Field(gt=0)
    beta: Decimal = Field(gt=0)
    beta_source: str = Field(default="manual", pattern="^(table3|table4|manual)$")


class UnitChemicalOut(UnitChemicalIn):
    id: str
    unit_id: str

    model_config = {"from_attributes": True}


class ComputeIn(BaseModel):
    exposed_population: int = Field(ge=0)


class CalculationOut(BaseModel):
    id: str
    seq: int
    s_value: Decimal
    r_value: Decimal
    alpha: Decimal
    exposed_population: int
    is_major_hazard: bool
    level: Optional[str] = None
    formula_version: str
    calculated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class CriticalQuantityOut(BaseModel):
    chemical_name: str
    alias: Optional[str] = None
    cas_no: Optional[str] = None
    critical_t: Optional[Decimal] = None
    critical_note: Optional[str] = None
    table_no: str
    source_page: Optional[int] = None

    model_config = {"from_attributes": True}


class EvidenceIn(BaseModel):
    """挂载法规依据的请求项。"""

    article_anchor: str = Field(min_length=1, max_length=200)
    regulation_id: Optional[str] = Field(default=None, max_length=64)
    relation: str = Field(default="依据", max_length=20)
    note: Optional[str] = None


class RecordIn(BaseModel):
    """档案与备案入参。未提供的字段保持原值。"""

    hazard_code: Optional[str] = Field(default=None, max_length=64)
    filing_status: Optional[str] = Field(default=None, max_length=20)
    filing_no: Optional[str] = Field(default=None, max_length=100)
    filing_date: Optional[date] = None
    chief_name: Optional[str] = None
    chief_post: Optional[str] = None
    chief_phone: Optional[str] = None
    tech_name: Optional[str] = None
    tech_post: Optional[str] = None
    tech_phone: Optional[str] = None
    oper_name: Optional[str] = None
    oper_post: Optional[str] = None
    oper_phone: Optional[str] = None
    attachments: Optional[dict] = None
    completeness: Optional[dict] = None


class RecordOut(BaseModel):
    id: str
    unit_id: str
    enterprise_id: str
    hazard_code: Optional[str] = None
    filing_status: str
    filing_no: Optional[str] = None
    filing_date: Optional[date] = None
    chief_name: Optional[str] = None
    chief_post: Optional[str] = None
    chief_phone: Optional[str] = None
    tech_name: Optional[str] = None
    tech_post: Optional[str] = None
    tech_phone: Optional[str] = None
    oper_name: Optional[str] = None
    oper_post: Optional[str] = None
    oper_phone: Optional[str] = None
    attachments: dict = Field(default_factory=dict)
    completeness: dict = Field(default_factory=dict)

    model_config = {"from_attributes": True}
